docs/madr/decisions/015-retry-policy.md

```diff
@@ -0,0 +1,223 @@
+      maxInterval: 1s
+    retryOn:
+      - cancelled
+      - deadline_exceeded
+      - internal
+      - resource_exhausted
+      - unavailable
+  tcp:
+    maxConnectAttempts: 3
+```
+
+Changes:
+
+1. Rename "http.retryOn.all_5xx" to "http.retryOn.5xx". Apparently it was a limitation of proto enums.
+2. Rename "retriableStatusCodes" to "additionalRetriableStatusCodes" 
+because this field provides additional status codes to whatever already specified in "retryOn".
+Today there is a validation if you set "retriableStatusCodes" you must provide "retriable_status_codes" to the "retryOn" list. 
+Due to a multi-level organisation of policies now it's impossible to validate.
+I think the best option is to add "retriable_status_codes" to "retryOn" automatically if "additionalRetriableStatusCodes" is specified. 
+3. New field "retriableHeaders" (or "retriableResponseHeaders"). Even though we have already had "retryOn.retriable_headers" 
+there was no way to provide a list of headers. 
+If this field is set we automatically add "retriable_headers" to "retryOn".
+4. New field "retriableRequestHeaders". Purely for the sake of symmetry, 
+if we allow retrying on response headers why not allow that on request headers. 
+
+#### Other retry-related Envoy parameters
+
+So far we didn't get requests from users to extend the policy with other Envoy parameters.
+But there are few additional things we can potentially support.
+
+##### per_try_idle_timeout
+
+Same as "idle_timeout" but enforced on each individual attempt.
+
+##### Host selection
+
+Envoy has "host predicates" and "priority predicates" to enforce an additional logic 
+when selecting a host for the attempt:
+
+* don't retry on previously retried hosts
+* don't retry on canary hosts
+* don't retry on hosts with specified metadata key/value
+* don't retry on previous priorities
+
+These predicates can be combined:
+
+```yaml
+hostSelection:
+  - predicate: OMIT_PREVIOUS_HOSTS
+  - predicate: OMIT_CANARY_HOSTS
+  - predicate: OMIT_HOSTS_WITH_TAGS
+    tags:
+      env: dev
+  - predicate: OMIT_PREVIOUS_PRIORITIES
+    updateFrequency: 2 # how often the priority load should be updated based on previously attempted priorities                                                                                             
+```
+
+Having configurable "hostSelection" implies exposing the "host_selection_retry_max_attempts" parameter as well.
+It allows configuring the maximum number of times host selection will be reattempted before giving up.
+
+##### Retry budged
+
+This parameter allows specifying a limit on concurrent retries in relation to the number of active requests.
+The problem is that "retry budged" is set on Envoy cluster 
+while other parameters are set on a "per route" or "per virtual host" basis.
+Even though today Kuma generates an Envoy cluster "per route" it might not be the case in the future.
+So I don't think having "retry budged" in MeshRetry policy is a good idea.
```

Comment:
Really nitpicky but I think when doing docs we need to consciously and consistently format, like using \`\` formatting when referring to fields/resources

Reply:
Right, fixed